app/src/main/java/protect/card_locker/importexport/CatimaImporter.java

```diff
@@ -368,22 +487,57 @@ private void importLoyaltyCard(Context context, SQLiteDatabase database, CSVReco
+    private void importGroupV2(SQLiteDatabase database, CSVRecord record) throws FormatException {
+        String name = CSVHelpers.extractString(DBHelper.LoyaltyCardDbGroups.ID, record, null);
+        DBHelper.insertGroup(database, name);
+    }
 
-        DBHelper.insertGroup(database, id);
+    /**
+     * Import a single group from the V3 scheme (database v16) into the database using the given
+     * session.
+     */
+    private Hashtable<Integer, String> importGroupV3(SQLiteDatabase database, List<CSVRecord> records) throws FormatException {
+        Hashtable<Integer,String> groupsTable = new Hashtable<>();
+        TreeMap<Integer, Group> sortedGroups = new TreeMap<Integer, Group>();
+        for (CSVRecord record : records) {
+            Integer id = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbGroups.ID, record, false);
+            String name = CSVHelpers.extractString(DBHelper.LoyaltyCardDbGroups.NAME, record, null);
+            Integer order = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbGroups.ORDER, record, false);
+            DBHelper.insertGroup(database, name);
+            sortedGroups.put(order, DBHelper.getGroupByName(database, name));
+            groupsTable.put(id, name);
+        }
+        List<Group> sortedGroupList = new ArrayList<>(sortedGroups.values());
+        DBHelper.reorderGroups(database, sortedGroupList);
+        return groupsTable;
     }
 
     /**
      * Import a single card to group mapping into the database using the given
      * session.
      */
-    private void importCardGroupMapping(SQLiteDatabase database, CSVRecord record) throws FormatException {
+    private void importCardGroupMappingV2(SQLiteDatabase database, CSVRecord record) throws FormatException {
         Integer cardId = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbIdsGroups.cardID, record, false);
-        String groupId = CSVHelpers.extractString(DBHelper.LoyaltyCardDbIdsGroups.groupID, record, null);
+        List<Group> cardGroups = DBHelper.getLoyaltyCardGroups(database, cardId);
+
+        String groupName = CSVHelpers.extractString(DBHelper.LoyaltyCardDbIdsGroups.groupID, record, null);
+        cardGroups.add(DBHelper.getGroupByName(database, groupName));
+
+        DBHelper.setLoyaltyCardGroups(database, cardId, cardGroups);
+    }
 
+    /**
+     * Import a single card to group mapping from V3 scheme (database v16) into the database using the given
+     * session.
+     */
+    private void importCardGroupMappingV3(SQLiteDatabase database, CSVRecord record, Hashtable<Integer,String> groupsTable) throws FormatException {
+        Integer cardId = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbIdsGroups.cardID, record, false);
         List<Group> cardGroups = DBHelper.getLoyaltyCardGroups(database, cardId);
+
+        Integer groupId = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbIdsGroups.groupID, record, false);
         cardGroups.add(DBHelper.getGroup(database, groupId));
```

Comment:
Finally figuring out why one of the tests is failing.

For one of the unit tests, this adds an entry to `cardGroups` with value `null`.

This is because DBHelper.getGroup() has 0 results for `groupId 1`.

I'm not quite sure yet why.

Reply:
Thanks a lot for figuring this out! Hmm I don't know why either... :thinking:

Reply:
Is maybe importCardGroupMappingV2 used instead of importCardGroupMappingV3, which defaults to null if no string found?

Reply:
Doesn't look like it, the failure happens in `importCardGroupMappingV3`. It's the `multipleCardsExportImportWithGroups` test failing.

```
E/Catima: Failed to import data
java.lang.NullPointerException
	at protect.card_locker.DBHelper.setLoyaltyCardGroups(DBHelper.java:641)
	at protect.card_locker.importexport.CatimaImporter.importCardGroupMappingV3(CatimaImporter.java:540)
	at protect.card_locker.importexport.CatimaImporter.parseV3CardGroups(CatimaImporter.java:361)
	at protect.card_locker.importexport.CatimaImporter.parseV3(CatimaImporter.java:211)
	at protect.card_locker.importexport.CatimaImporter.importCSV(CatimaImporter.java:88)
	at protect.card_locker.importexport.CatimaImporter.importData(CatimaImporter.java:59)
	at protect.card_locker.importexport.MultiFormatImporter.importData(MultiFormatImporter.java:47)
	at protect.card_locker.ImportExportTest.multipleCardsExportImportWithGroups(ImportExportTest.java:468)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:580)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$2(SandboxTestRunner.java:287)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:99)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```